### PR TITLE
to_pil - don't rescale if int and in range 0-255

### LINKED ISSUE #22147: OneFormerProcessor、MaskFormerImageProcessor will cause errors if segmentation_maps only have elements 0 and 1
### System Info

transformers-4.26.0  do not have this bug
but transformers-4.27.0.dev0 has.

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```python
from transformers import OneFormerProcessor, OneFormerForUniversalSegmentation, OneFormerImageProcessor, OneFormerConfig
from transformers import Mask2FormerImageProcessor, Mask2FormerForUniversalSegmentation
from PIL import Image
import requests
import torch
import numpy as np
import matplotlib

processor = OneFormerProcessor.from_pretrained("shi-labs/oneformer_ade20k_swin_tiny",num_text=134,do_reduce_labels=True,)

image_np=np.random.randint(0,255,(3,512,512))

#segmentation_maps only have elements 0 and 1
segmentation_maps = torch.randint(0, 2, (image_np.shape[1], image_np.shape[2]), dtype=torch.long)

inst2class={1: 4}

raw_inputs=processor.image_processor([image_np],
                   task_inputs=["panoptic"],
                   segmentation_maps=[segmentation_maps],
                   return_tensors="pt",
                   instance_id_to_semantic_id=inst2class,
                   do_reduce_labels=True,
                   ignore_index=None)

```
#ERROR
```
E:\condaenv\yaogan\lib\site-packages\transformers\models\oneformer\image_processing_oneformer.py:419: FutureWarning: The `reduce_labels` argument is deprecated and will be removed in v4.27. Please use `do_reduce_labels` instead.
  warnings.warn(
Traceback (most recent call last):
  File "E:\condaenv\yaogan\lib\site-packages\IPython\core\interactiveshell.py", line 3460, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-ed9733992fe8>", line 23, in <module>
    raw_inputs=processor.image_processor([image_np],
  File "E:\condaenv\yaogan\lib\site-packages\transformers\models\oneformer\image_processing_oneformer.py", line 524, in __call__
    return self.preprocess(images, task_inputs=task_inputs, segmentation_maps=segmentation_maps, **kwargs)
  File "E:\condaenv\yaogan\lib\site-packages\transformers\models\oneformer\image_processing_oneformer.py", line 708, in preprocess
    encoded_inputs = self.encode_inputs(
  File "E:\condaenv\yaogan\lib\site-packages\transformers\models\oneformer\image_processing_oneformer.py", line 962, in encode_inputs
    masks, classes = self.convert_segmentation_map_to_binary_masks(
  File "E:\condaenv\yaogan\lib\site-packages\transformers\models\oneformer\image_processing_oneformer.py", line 516, in convert_segmentation_map_to_binary_masks
    return convert_segmentation_map_to_binary_masks(
  File "E:\condaenv\yaogan\lib\site-packages\transformers\models\oneformer\image_processing_oneformer.py", line 288, in convert_segmentation_map_to_binary_masks
    class_id = instance_id_to_semantic_id[label + 1 if reduce_labels else label]
KeyError: 255
```
This bug is caused by a **resize** function of OneFormerProcessor, which convert segmentation_maps to PIL.Image and then convert to np.ndarray. After **resize**, segmentation_maps have elements 0 and 255, so the bug arise. 

### Expected behavior

fix this bug before release 4.27.0 as stable version
transformers-4.26.0  do not have this bug

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

#21969 Introduced a bug where binary masks would have their values rescaled to between 0-255. 

This was because of [this part](https://github.com/huggingface/transformers/blob/4063fd9cba6b72ebfd5c663a307ab9d5ff1a153d/src/transformers/image_transforms.py#L161) of the logic check. The original assumption was that inputs with their values between 0-1 would be rescaled images with float pixels. However, binary masks aren't and shouldn't be rescaled. 

We now check first if the input is of type uint8. Then check if any precision is lost when converting to int and that the int values are in the valid range 0-255 and finally if float values are between 0-1. 

Fixes #22147

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks for the fix and the new tests!

File: tests/test_image_transforms.py [hunk 1]
```diff
@@ -101,6 +101,33 @@ def test_to_pil_image_from_float(self, name, image_shape, dtype):
         with self.assertRaises(ValueError):
             to_pil_image(image)
 
+        # Make sure binary mask remains a binary mask
+        image = np.random.randint(0, 2, image_shape).astype(np.uint8)
+        pil_image = to_pil_image(image)
+        self.assertIsInstance(pil_image, PIL.Image.Image)
+        self.assertEqual(pil_image.size, (5, 4))
```
Comment (ydshieh) [inline tests/test_image_transforms.py:108]: I feel here we miss the objective of the check: `binary mask remains a binary mask` ..?

Comment (ydshieh) [reply tests/test_image_transforms.py:108]: it also looks this check is implemented in `test_to_pil_image_from_mask` below, I am not very sure the reason of this new block here.

**Review by ydshieh — APPROVED** [review]
Very nice. Just 2 nit questions :-).

Thank you @amyeroberts

File: tests/test_image_transforms.py (hunk 1 above)
Comment (amyeroberts) [reply tests/test_image_transforms.py:108]: You're right. I originally put the check here but then moved below as this method is `test_to_pil_image_from_float` so didn't want to use int inputs. Forgot to delete. I'll remove 👍

Comment (amyeroberts) [inline tests/test_image_transforms.py:108]:
```suggestion
```

File: tests/test_image_transforms.py [hunk 2]
```diff
@@ -101,6 +101,28 @@ def test_to_pil_image_from_float(self, name, image_shape, dtype):
         with self.assertRaises(ValueError):
             to_pil_image(image)
 
+
```
Comment (amyeroberts) [inline tests/test_image_transforms.py:104]:
```suggestion
```

**Pull request merged**